oxd-server/src/main/java/org/gluu/oxd/server/service/KeyGeneratorService.java

```diff
@@ -0,0 +1,166 @@
+import org.gluu.oxd.common.ErrorResponseCode;
+import org.gluu.oxd.server.HttpException;
+import org.gluu.oxd.server.OxdServerConfiguration;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.KeyStoreException;
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+import java.util.List;
+
+import static org.gluu.oxauth.model.jwk.JWKParameter.*;
+
+public class KeyGeneratorService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KeyGeneratorService.class);
+
+    private OxdServerConfiguration configuration;
+
+    private AbstractCryptoProvider cryptoProvider;
+
+    private JSONWebKeySet keys;
+
+    @Inject
+    public KeyGeneratorService(OxdServerConfiguration configuration) {
+        this.configuration = configuration;
+        this.keys = new JSONWebKeySet();
+        try {
+            this.cryptoProvider = new OxAuthCryptoProvider(configuration.getCryptProviderKeyStorePath(), configuration.getCryptProviderKeyStorePassword(), configuration.getCryptProviderDnName());
+        } catch (Exception e) {
+            LOG.error("Failed to create CryptoProvider.", e);
+            throw new RuntimeException("Failed to create CryptoProvider.", e);
+        }
+    }
+
+    public void load() {
+        List<Algorithm> signatureAlgorithms = Lists.newArrayList(Algorithm.RS256, Algorithm.RS384, Algorithm.RS512, Algorithm.ES256,
+                Algorithm.ES384, Algorithm.ES512, Algorithm.PS256, Algorithm.PS384, Algorithm.PS512);
+
+        List<Algorithm> encryptionAlgorithms = Lists.newArrayList(Algorithm.RSA1_5, Algorithm.RSA_OAEP);
+
+        try {
+            if (configuration.getEnableJwksGeneration()) {
+                this.keys = generateKeys(signatureAlgorithms, encryptionAlgorithms, configuration.getJwksExpirationInDays(), configuration.getJwksExpirationInHours());
+            }
+        } catch (Exception e) {
+            LOG.error("Failed to generate json web keys.", e);
+            throw new RuntimeException("Failed to generate json web keys.", e);
+        }
+    }
+
+    private JSONWebKeySet generateKeys(List<Algorithm> signatureAlgorithms,
+                                       List<Algorithm> encryptionAlgorithms, int expiration, int expiration_hours) throws Exception, JSONException {
+        JSONWebKeySet jwks = new JSONWebKeySet();
+
+        Calendar calendar = new GregorianCalendar();
+        calendar.add(Calendar.DATE, expiration);
+        calendar.add(Calendar.HOUR, expiration_hours);
+
+        for (Algorithm algorithm : signatureAlgorithms) {
+            try {
+                SignatureAlgorithm signatureAlgorithm = SignatureAlgorithm.fromString(algorithm.name());
+                JSONObject result = this.cryptoProvider.generateKey(algorithm, calendar.getTimeInMillis(), Use.SIGNATURE);
+
+                JSONWebKey key = new JSONWebKey();
```

Comment:
Would this work instead of manual re-creation?
`org.gluu.oxauth.model.jwk.JSONWebKey#fromJSONObject`

Reply:
Done